JWT: Allow conventional bearer token in Authorization header

nrwiersma opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR fixes the issue that JWT tokens in the `Authorization` header are incorrectly declined by the API Key auth mechanism. It also allows the handling of JWT tokens with a `Bearer` prefix in a backwards compatible manner.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #54244

**Special notes for your reviewer**:

This was the only non-breaking way I have found to allow the desired behaviour. To lower the chance that the API Key auth will incorrectly allow the auth request to fall through, the test for weather it is a JWT token in the `Authorization` header has been made reasonable robust while attempting not to have too much of a performance impact. If more robustness is required at the cost of performance, there are further tests that can be performed on the token.

nrwiersma committed:
fix: allow JWT to accept standard bearer token

nrwiersma committed:
fix: linter issues

nrwiersma committed:
fix: linter gosec false positive

sakjur commented:
Are you talking about a formalized standard when you write standard in the title? RFC7519, RFC7797 and RFC8725 makes no reference to using the Bearer scheme to carry JWTs. If you’re talking about a convention or de facto standard, please remove the word standard from the title, and if you’re talking about a formal standard, please link it so that we can determine whether it’s a standard we want to support.

@Jguer what would you think about placing the JWT handler before the API key handler and have it fall through if it is configured to listen to the Authorization header and doesn’t recognize the token as a JWT? That would remove the need for the API key handler to need to know anything about JWTs.

Jguer commented:
@sakjur That's a good idea, specially as the config flag can be used to easily fall-through. @nrwiersma would you be available for making these changes?

nrwiersma commented:
@Jguer Sure, will make the change. To be clear, we want the same checks that I have already implemented, just moved into JWT? Or are there other checks you would want performed? I would assume that once the JWT verification starts (and fails) and error should still be thrown.

nrwiersma committed:
fix: refactor logic into JWT handler

sakjur commented:
Yes, I believe you could do that by moving the JWT detection logic you have for the API key to the JWT handler, and then if the configured header is `Authorization`, return false if the scheme isn't `Bearer` or the JWT isn't a plausible JWT (dot-detection should work for that, I don't think we generate API keys with dots in them), otherwise strip away the Bearer part of the header and go on to validate the JWT 🤔

(or something to the same effect)

nrwiersma committed:
fix: move bearer trimming earlier

Jguer reviewed (approved):
Tested it :+1: working fine on API Key/SAT passthrough while maintaining JWT functionality. Thanks for your contribution again @nrwiersma
